File: udon_decompiler/analysis/transform/passes/control_flow_simplification.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, Iterator, Optional

from udon_decompiler.analysis.ir.nodes import (
    IRBlock,
    IRBlockContainer,
    IRFunction,
    IRHighLevelSwitch,
    IRIf,
    IRJump,
    IRLeave,
    IRReturn,
    IRStatement,
    IRSwitch,
)
from udon_decompiler.analysis.transform.pass_base import (
    ITransform,
    TransformContext,
)


@dataclass
class _SimplifiedJump:
    statement: IRStatement
    changed: bool


class ControlFlowSimplification(ITransform):
    """
    Early CFG simplification on current IR.

    Current scope:
    - simplify branch-to-branch chains
    - replace branch-to-leave with direct leave
    - combine basic blocks where safe
    - remove dead empty blocks
    """

    def run(self, function: IRFunction, context: TransformContext) -> None:
        # Run until fixed point
        for _ in range(32):
            changed = False
            if self._simplify_branch_chains(function, context):
                changed = True
            if self._cleanup_empty_blocks(function, context):
                changed = True
            if not changed:
                break

    def _simplify_branch_chains(
        self,
        function: IRFunction,
        context: TransformContext,
    ) -> bool:
        changed = False
        for container in self._iter_containers(function.body):
            for block in container.blocks:
                for index, statement in enumerate(list(block.statements)):
                    updated, stmt_changed = self._rewrite_statement(statement, context)
                    if stmt_changed:
                        block.statements[index] = updated
                        changed = True
        return changed

    def _rewrite_statement(
        self,
        statement: IRStatement,
        context: TransformContext,
    ) -> tuple[IRStatement, bool]:
        if isinstance(statement, IRJump):
            simplified = self._simplify_jump(statement, context)
            return simplified.statement, simplified.changed

        if isinstance(statement, IRIf):
            true_stmt, true_changed = self._rewrite_statement(
                statement.true_statement,
                context,
            )
            false_changed = False
            false_stmt = statement.false_statement
            if false_stmt is not None:
                false_stmt, false_changed = self._rewrite_statement(
                    false_stmt,
                    context,
                )
            statement.true_statement = true_stmt
            statement.false_statement = false_stmt
            return statement, (true_changed or false_changed)

        if isinstance(statement, IRBlock):
            changed = False
            for index, nested in enumerate(list(statement.statements)):
                updated, nested_changed = self._rewrite_statement(nested, context)
                if nested_changed:
                    statement.statements[index] = updated
                    changed = True
            return statement, changed

        if isinstance(statement, IRSwitch):
            changed = False
            for case_value, target in list(statement.cases.items()):
                resolved_target, target_changed = self._resolve_branch_target(target)
                if target_changed:
                    statement.cases[case_value] = resolved_target
                    changed = True
            if statement.default_target is not None:
                resolved_default, default_changed = self._resolve_branch_target(
                    statement.default_target
                )
                if default_changed:
                    statement.default_target = resolved_default
                    changed = True
            return statement, changed

        if isinstance(statement, IRHighLevelSwitch):
            changed = False
            for section in statement.sections:
                updated_body, body_changed = self._rewrite_statement(
                    section.body,
                    context,
                )
                if body_changed:
                    section.body = updated_body  # type: ignore[assignment]
                    changed = True
            return statement, changed

        if isinstance(statement, IRBlockContainer):
            changed = False
            for block in statement.blocks:
                for index, nested in enumerate(list(block.statements)):
                    updated, nested_changed = self._rewrite_statement(nested, context)
                    if nested_changed:
                        block.statements[index] = updated
                        changed = True
            return statement, changed

        return statement, False

    def _simplify_jump(
        self,
        jump: IRJump,
        context: TransformContext,
    ) -> _SimplifiedJump:
        original_target = jump.target
        target, chain_simplified = self._resolve_branch_target(original_target)
        if chain_simplified:
            context.step("Simplify branch to branch")

        if len(target.statements) == 1 and isinstance(target.statements[0], IRLeave):
            context.step("Replace branch to leave with leave")
            leave = target.statements[0]
            return _SimplifiedJump(
                statement=IRLeave(target_container=leave.target_container),
                changed=True,
            )

        if chain_simplified and target is not original_target:
            return _SimplifiedJump(statement=IRJump(target=target), changed=True)

        return _SimplifiedJump(statement=jump, changed=chain_simplified)

    def _resolve_branch_target(self, target: IRBlock) -> tuple[IRBlock, bool]:
        visited: set[IRBlock] = set()
        changed = False
        current = target

        while (
            len(current.statements) == 1
            and isinstance(current.statements[0], IRJump)
            and current not in visited
        ):
            visited.add(current)
            current = current.statements[0].target
            changed = True

        return current, changed

    def _cleanup_empty_blocks(
        self,
        function: IRFunction,
        context: TransformContext,
    ) -> bool:
        changed = False

        for container in self._iter_containers(function.body):
            while True:
                incoming = self._compute_global_incoming_counts(function.body)
                combined = False
                for block in list(container.blocks):
                    if self._combine_block_with_next(
                        container=container,
                        block=block,
                        incoming=incoming,
                        context=context,
                    ):
                        changed = True
                        combined = True
                        break
                if not combined:
                    break

            incoming = self._compute_global_incoming_counts(function.body)
            entry = container.entry_block
            kept_blocks: list[IRBlock] = []
            removed = False
            for block in container.blocks:
                if block is entry:
                    kept_blocks.append(block)
                    continue
                if incoming.get(block, 0) == 0:
                    removed = True
                    continue
                kept_blocks.append(block)
            if removed:
                context.step("Remove dead blocks")
                container.blocks = kept_blocks
                changed = True

        return changed

    def _combine_block_with_next(
        self,
        container: IRBlockContainer,
        block: IRBlock,
        incoming: Dict[IRBlock, int],
        context: TransformContext,
    ) -> bool:
        if block not in container.blocks:
            return False
        if not block.statements:
            return False

        if len(block.statements) > 1 and self._statement_may_branch(
            block.statements[-2]
        ):
            return False

        terminal = block.statements[-1]
        if not isinstance(terminal, IRJump):
            return False

        target_block = terminal.target
        if target_block not in container.blocks:
            return False
        if target_block is block:
            return False
        if incoming.get(target_block, 0) != 1:
            return False

        context.step("Combine block with successor")
        block.statements.pop()
        block.statements.extend(target_block.statements)
        target_block.statements.clear()
        return True

    @staticmethod
    def _statement_may_branch(statement: IRStatement) -> bool:
        if isinstance(statement, (IRIf, IRJump, IRLeave, IRReturn, IRSwitch)):
            return True
        if isinstance(statement, IRHighLevelSwitch):
            return True
        if isinstance(statement, IRBlockContainer):
            return True
        return False

    def _compute_global_incoming_counts(
        self,
        body: IRBlockContainer,
    ) -> Dict[IRBlock, int]:
        counts: Dict[IRBlock, int] = {}
        for container in self._iter_containers(body):
            for block in container.blocks:
                counts.setdefault(block, 0)

        for container in self._iter_containers(body):
            for block in container.blocks:
                for statement in block.statements:
                    self._count_statement_targets(statement, counts)
        return counts

    def _count_statement_targets(
        self,
        statement: IRStatement,
        counts: Dict[IRBlock, int],
    ) -> None:
        if isinstance(statement, IRJump):
            counts[statement.target] = counts.get(statement.target, 0) + 1
            return

        if isinstance(statement, IRIf):
            self._count_statement_targets(statement.true_statement, counts)
            if statement.false_statement is not None:
                self._count_statement_targets(statement.false_statement, counts)
            return

        if isinstance(statement, IRSwitch):
            for target in statement.cases.values():
                counts[target] = counts.get(target, 0) + 1
            if statement.default_target is not None:
                target = statement.default_target
                counts[target] = counts.get(target, 0) + 1
            return

        if isinstance(statement, IRHighLevelSwitch):
            for section in statement.sections:
                self._count_statement_targets(section.body, counts)
            return

        if isinstance(statement, IRBlock):
            for nested in statement.statements:
                self._count_statement_targets(nested, counts)
            return

        if isinstance(statement, IRBlockContainer):
            for block in statement.blocks:
                for nested in block.statements:
                    self._count_statement_targets(nested, counts)

    def _iter_containers(
        self,
        root: IRBlockContainer,
    ) -> Iterator[IRBlockContainer]:
        visited: set[int] = set()
        stack: list[IRBlockContainer] = [root]

        while stack:
            container = stack.pop()
            key = id(container)
            if key in visited:
                continue
            visited.add(key)
            yield container

            for block in container.blocks:
                for statement in block.statements:
                    stack.extend(self._nested_containers_in_statement(statement))

    def _nested_containers_in_statement(
        self,
        statement: IRStatement,
    ) -> list[IRBlockContainer]:
        if isinstance(statement, IRBlockContainer):
            result = [statement]
            for block in statement.blocks:
                for nested in block.statements:
                    result.extend(self._nested_containers_in_statement(nested))
            return result

        if isinstance(statement, IRIf):
            result = self._nested_containers_in_statement(statement.true_statement)
            if statement.false_statement is not None:
                result.extend(
                    self._nested_containers_in_statement(statement.false_statement)
                )
            return result

        if isinstance(statement, IRBlock):
            result: list[IRBlockContainer] = []
            for nested in statement.statements:
                result.extend(self._nested_containers_in_statement(nested))
            return result

        if isinstance(statement, IRHighLevelSwitch):
            result: list[IRBlockContainer] = []
            for section in statement.sections:
                result.extend(self._nested_containers_in_statement(section.body))
            return result

        return []

File: src/decompiler/transform/passes/control_flow_simplification.rs
use std::collections::{HashMap, HashSet};

use crate::decompiler::Result;
use crate::decompiler::ir::{
    IrBlock, IrBlockContainer, IrFunction, IrHighLevelDoWhile, IrHighLevelSwitch, IrHighLevelWhile,
    IrIf, IrJump, IrLeave, IrReturn, IrStatement, IrSwitch,
};
use crate::decompiler::transform::pass_base::{ITransform, TransformContext};

pub struct ControlFlowSimplification;

impl ITransform for ControlFlowSimplification {
    fn run(
        &self,
        function: &mut IrFunction,
        _context: &mut TransformContext<'_, '_>,
    ) -> Result<()> {
        for _ in 0..32 {
            let mut changed = false;
            if simplify_branch_chains(function) {
                changed = true;
            }
            if cleanup_empty_blocks(function) {
                changed = true;
            }
            if !changed {
                break;
            }
        }
        Ok(())
    }
}

fn simplify_branch_chains(function: &mut IrFunction) -> bool {
    let mut address_to_block = HashMap::<u32, IrBlock>::new();
    collect_blocks_map(&function.body, &mut address_to_block);
    rewrite_container_statements(&mut function.body, &address_to_block)
}

fn rewrite_container_statements(
    container: &mut IrBlockContainer,
    address_to_block: &HashMap<u32, IrBlock>,
) -> bool {
    let mut changed = false;
    for block in &mut container.blocks {
        for statement in &mut block.statements {
            if rewrite_statement(statement, address_to_block) {
                changed = true;
            }
        }
    }
    changed
}

fn rewrite_statement(
    statement: &mut IrStatement,
    address_to_block: &HashMap<u32, IrBlock>,
) -> bool {
    match statement {
        IrStatement::Jump(IrJump { target_address }) => {
            let original = *target_address;
            let (resolved, chain_changed) = resolve_branch_target(original, address_to_block);

            if let Some(target_block) = address_to_block.get(&resolved)
                && target_block.statements.len() == 1
                && let Some(IrStatement::Leave(IrLeave {
                    target_container_id,
                })) = target_block.statements.first()
            {
                *statement = IrStatement::Leave(IrLeave {
                    target_container_id: *target_container_id,
                });
                return true;
            }

            if chain_changed {
                *target_address = resolved;
                return true;
            }

            false
        }
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            let mut changed = rewrite_statement(true_statement, address_to_block);
            if let Some(false_statement) = false_statement.as_mut()
                && rewrite_statement(false_statement, address_to_block)
            {
                changed = true;
            }
            changed
        }
        IrStatement::Switch(IrSwitch {
            cases,
            default_target,
            ..
        }) => {
            let mut changed = false;
            for target in cases.values_mut() {
                let (resolved, target_changed) = resolve_branch_target(*target, address_to_block);
                if target_changed {
                    *target = resolved;
                    changed = true;
                }
            }
            if let Some(default_target) = default_target.as_mut() {
                let (resolved, default_changed) =
                    resolve_branch_target(*default_target, address_to_block);
                if default_changed {
                    *default_target = resolved;
                    changed = true;
                }
            }
            changed
        }
        IrStatement::Block(block) => {
            let mut changed = false;
            for nested in &mut block.statements {
                if rewrite_statement(nested, address_to_block) {
                    changed = true;
                }
            }
            changed
        }
        IrStatement::BlockContainer(container) => {
            rewrite_container_statements(container, address_to_block)
        }
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            let mut changed = false;
            for section in sections {
                for nested in &mut section.body.statements {
                    if rewrite_statement(nested, address_to_block) {
                        changed = true;
                    }
                }
            }
            changed
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => {
            rewrite_container_statements(body, address_to_block)
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => {
            rewrite_container_statements(body, address_to_block)
        }
        _ => false,
    }
}

fn resolve_branch_target(
    target_address: u32,
    address_to_block: &HashMap<u32, IrBlock>,
) -> (u32, bool) {
    let mut current = target_address;
    let mut changed = false;
    let mut visited = HashSet::<u32>::new();

    while visited.insert(current) {
        let Some(block) = address_to_block.get(&current) else {
            break;
        };
        if block.statements.len() != 1 {
            break;
        }
        let Some(IrStatement::Jump(IrJump { target_address })) = block.statements.first() else {
            break;
        };
        current = *target_address;
        changed = true;
    }

    (current, changed)
}

fn cleanup_empty_blocks(function: &mut IrFunction) -> bool {
    let mut changed = false;

    loop {
        let incoming = compute_global_incoming_counts(function);
        if !combine_once_in_container(&mut function.body, &incoming) {
            break;
        }
        changed = true;
    }

    let incoming = compute_global_incoming_counts(function);
    if remove_dead_blocks_in_container(&mut function.body, &incoming) {
        changed = true;
    }

    changed
}

fn combine_once_in_container(
    container: &mut IrBlockContainer,
    incoming: &HashMap<u32, usize>,
) -> bool {
    for idx in 0..container.blocks.len() {
        let Some(block) = container.blocks.get(idx) else {
            continue;
        };
        if block.statements.is_empty() {
            continue;
        }
        if block.statements.len() > 1
            && block
                .statements
                .get(block.statements.len() - 2)
                .is_some_and(statement_may_branch)
        {
            continue;
        }

        let Some(IrStatement::Jump(IrJump { target_address })) = block.statements.last() else {
            continue;
        };
        let Some(target_idx) = container
            .blocks
            .iter()
            .position(|x| x.start_address == *target_address)
        else {
            continue;
        };
        if target_idx == idx {
            continue;
        }
        if incoming.get(target_address).copied().unwrap_or(0) != 1 {
            continue;
        }

        let mut merged = container.blocks[idx].statements.clone();
        let _ = merged.pop();
        merged.extend(container.blocks[target_idx].statements.clone());
        container.blocks[idx].statements = merged;
        container.blocks[target_idx].statements.clear();
        return true;
    }

    for block in &mut container.blocks {
        for statement in &mut block.statements {
            if combine_once_in_statement(statement, incoming) {
                return true;
            }
        }
    }

    false
}

fn combine_once_in_statement(statement: &mut IrStatement, incoming: &HashMap<u32, usize>) -> bool {
    match statement {
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            if combine_once_in_statement(true_statement, incoming) {
                return true;
            }
            if let Some(false_statement) = false_statement.as_mut()
                && combine_once_in_statement(false_statement, incoming)
            {
                return true;
            }
        }
        IrStatement::Block(block) => {
            for nested in &mut block.statements {
                if combine_once_in_statement(nested, incoming) {
                    return true;
                }
            }
        }
        IrStatement::BlockContainer(container) => {
            if combine_once_in_container(container, incoming) {
                return true;
            }
        }
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            for section in sections {
                for nested in &mut section.body.statements {
                    if combine_once_in_statement(nested, incoming) {
                        return true;
                    }
                }
            }
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => {
            if combine_once_in_container(body, incoming) {
                return true;
            }
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => {
            if combine_once_in_container(body, incoming) {
                return true;
            }
        }
        _ => {}
    }
    false
}

fn remove_dead_blocks_in_container(
    container: &mut IrBlockContainer,
    incoming: &HashMap<u32, usize>,
) -> bool {
    let mut changed = false;
    let mut kept = Vec::<IrBlock>::new();
    for (index, block) in container.blocks.iter().cloned().enumerate() {
        if index == 0 || incoming.get(&block.start_address).copied().unwrap_or(0) > 0 {
            kept.push(block);
        } else {
            changed = true;
        }
    }
    container.blocks = kept;

    for block in &mut container.blocks {
        for statement in &mut block.statements {
            if remove_dead_blocks_in_statement(statement, incoming) {
                changed = true;
            }
        }
    }

    changed
}

fn remove_dead_blocks_in_statement(
    statement: &mut IrStatement,
    incoming: &HashMap<u32, usize>,
) -> bool {
    match statement {
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            let mut changed = remove_dead_blocks_in_statement(true_statement, incoming);
            if let Some(false_statement) = false_statement.as_mut()
                && remove_dead_blocks_in_statement(false_statement, incoming)
            {
                changed = true;
            }
            changed
        }
        IrStatement::Block(block) => {
            let mut changed = false;
            for nested in &mut block.statements {
                if remove_dead_blocks_in_statement(nested, incoming) {
                    changed = true;
                }
            }
            changed
        }
        IrStatement::BlockContainer(container) => {
            remove_dead_blocks_in_container(container, incoming)
        }
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            let mut changed = false;
            for section in sections {
                for nested in &mut section.body.statements {
                    if remove_dead_blocks_in_statement(nested, incoming) {
                        changed = true;
                    }
                }
            }
            changed
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => {
            remove_dead_blocks_in_container(body, incoming)
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => {
            remove_dead_blocks_in_container(body, incoming)
        }
        _ => false,
    }
}

fn statement_may_branch(statement: &IrStatement) -> bool {
    matches!(
        statement,
        IrStatement::If(_)
            | IrStatement::Jump(_)
            | IrStatement::Leave(_)
            | IrStatement::Return(IrReturn)
            | IrStatement::Switch(_)
            | IrStatement::HighLevelSwitch(_)
            | IrStatement::BlockContainer(_)
    )
}

fn compute_global_incoming_counts(function: &IrFunction) -> HashMap<u32, usize> {
    let mut counts = HashMap::<u32, usize>::new();
    collect_block_addresses(&function.body, &mut counts);
    count_targets_in_container(&function.body, &mut counts);
    counts
}

fn collect_block_addresses(container: &IrBlockContainer, counts: &mut HashMap<u32, usize>) {
    for block in &container.blocks {
        counts.entry(block.start_address).or_insert(0);
        for statement in &block.statements {
            collect_block_addresses_from_statement(statement, counts);
        }
    }
}

fn collect_block_addresses_from_statement(
    statement: &IrStatement,
    counts: &mut HashMap<u32, usize>,
) {
    match statement {
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            collect_block_addresses_from_statement(true_statement, counts);
            if let Some(false_statement) = false_statement.as_ref() {
                collect_block_addresses_from_statement(false_statement, counts);
            }
        }
        IrStatement::Block(block) => {
            for nested in &block.statements {
                collect_block_addresses_from_statement(nested, counts);
            }
        }
        IrStatement::BlockContainer(container) => collect_block_addresses(container, counts),
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            for section in sections {
                for nested in &section.body.statements {
                    collect_block_addresses_from_statement(nested, counts);
                }
            }
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => {
            collect_block_addresses(body, counts);
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => {
            collect_block_addresses(body, counts);
        }
        _ => {}
    }
}

fn count_targets_in_container(container: &IrBlockContainer, counts: &mut HashMap<u32, usize>) {
    for block in &container.blocks {
        for statement in &block.statements {
            count_targets_in_statement(statement, counts);
        }
    }
}

fn count_targets_in_statement(statement: &IrStatement, counts: &mut HashMap<u32, usize>) {
    match statement {
        IrStatement::Jump(IrJump { target_address }) => {
            *counts.entry(*target_address).or_insert(0) += 1;
        }
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            count_targets_in_statement(true_statement, counts);
            if let Some(false_statement) = false_statement.as_ref() {
                count_targets_in_statement(false_statement, counts);
            }
        }
        IrStatement::Switch(IrSwitch {
            cases,
            default_target,
            ..
        }) => {
            for target in cases.values() {
                *counts.entry(*target).or_insert(0) += 1;
            }
            if let Some(default_target) = default_target {
                *counts.entry(*default_target).or_insert(0) += 1;
            }
        }
        IrStatement::Block(block) => {
            for nested in &block.statements {
                count_targets_in_statement(nested, counts);
            }
        }
        IrStatement::BlockContainer(container) => {
            count_targets_in_container(container, counts);
        }
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            for section in sections {
                for nested in &section.body.statements {
                    count_targets_in_statement(nested, counts);
                }
            }
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => {
            count_targets_in_container(body, counts);
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => {
            count_targets_in_container(body, counts);
        }
        _ => {}
    }
}

fn collect_blocks_map(container: &IrBlockContainer, out: &mut HashMap<u32, IrBlock>) {
    for block in &container.blocks {
        out.insert(block.start_address, block.clone());
        for statement in &block.statements {
            collect_blocks_map_from_statement(statement, out);
        }
    }
}

fn collect_blocks_map_from_statement(statement: &IrStatement, out: &mut HashMap<u32, IrBlock>) {
    match statement {
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            collect_blocks_map_from_statement(true_statement, out);
            if let Some(false_statement) = false_statement.as_ref() {
                collect_blocks_map_from_statement(false_statement, out);
            }
        }
        IrStatement::Block(block) => {
            for nested in &block.statements {
                collect_blocks_map_from_statement(nested, out);
            }
        }
        IrStatement::BlockContainer(container) => collect_blocks_map(container, out),
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            for section in sections {
                for nested in &section.body.statements {
                    collect_blocks_map_from_statement(nested, out);
                }
            }
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => {
            collect_blocks_map(body, out);
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => {
            collect_blocks_map(body, out);
        }
        _ => {}
    }
}
